[5.9] Add ability to get normalized response from gate: Gate::access(...)

garygreen opened the pull request:
Following on from the recent addition of specifying a reason `code` when denying policies this PR adds a method onto the `Gate` which allows you to get the `Response` object for a given ability check.

### Why is this useful?

A good use case for this is anytime you want to check if an ability is not granted and understand *why*, without it throwing an exception.

### How is this solved at the moment in userland?

This can be solved by using `authorize` and catching the `AuthorizationException` but this quickly becomes unmaintainable if you need to check lots of permissions (see serialization frontend example).

```php
// Image if you had to check lots of user permissions
// and pass the deny messages and reasons onto the view... could get messy.
try {
    Gate::authorize('delete', $user);
    $allowed = true;
} catch (AuthorizationException $e) {
    $allowed = false;
    $denyCode = $e->getCode();
}
```

### What this PR does

Instead of the above, with this PR you can now do:

```php
$response = Gate::response('delete', $user);
$response->allowed(); // bool
$response->denied(); // bool
$response->message(); // Message from the policy, both allow and deny could send a message.
$response->code(); // The code from the policy, both allow and deny could send a code/reason of some kind.
```

This allows for more complex serialization of your permissions for example if you wanted to pass them onto your frontend for use in Javascript:

```php
return view('users.index', [
    'permissions' => $users->mapWithKeys(function($user) {
        return [
            "{$user->getKey()}.edit" => Gate::response('edit', $user),
            "{$user->getKey()}.delete" => Gate::response('delete', $user),
        ];
    }),
]);
```

```html
// In your view:
<script>
const permissions = '{{ $permissions->toJson() }}';
</script>
```

The above will serialize by default like:

```json
{
    "3.edit": {
        "allowed": true,
        "message": null,
        "code": null
    },
    "3.delete": {
        "allowed": false,
        "message": "You are not allowed to delete this user because they are an admin.",
        "code": "user_admin"
    }
}
```

I believe this PR is the final key needed to have a comprehensive workflow related to handling permissions. @JosephSilber mentioned on twitter he would also like to see this kind of normalization for an access response.

garygreen committed:
Add ability to get normalized response from gate: Gate::access(...)

garygreen commented:
ping @d3jn you may find this useful

JosephSilber commented:
The `check` and `authorize` methods need to also handle a failed response properly.

garygreen commented:
> The check and authorize methods need to also handle a failed response properly.

Could you clarify? The implementation of `check` and `authorize` I believe should remain unchanged otherwise it would be a pretty big breaking change..

garygreen commented:
@JosephSilber just pushed a commit, I believe this fixes the `check` and `authorize` method

driesvints commented:
@garygreen there still seems to be a failing test.

garygreen commented:
Thanks @driesvints have updated and added some more tests, also done a bit of refactoring.

taylorotwell commented:
Pretty good with the overall idea. Trying to think of a better name than `response` at the moment.

danilopinotti commented:
> Pretty good with the overall idea. Trying to think of a better name than `response` at the moment.

Maybe 'resolve'?

garygreen commented:
What about `access` ?

```php
Gate::access('send', $message)->allowed();
Gate::access('send', $message)->message();
```

This could also work in your controllers if you assign intermediary variable:

```php
$access = Gate::access('send', $message);

if ($access->code == 'vip_required') {
   return redirect()->route('purchase.vip')->withError($access->message());
}
```

garygreen commented:
I've gone ahead and made a few changes:

* Renamed `response` to `access`: usage is now `Gate::access()`
* Renamed `$response->throwIfNeeded()` to simply `$response->authorize()` - this better aligns with `Gate::authorize()` usage.

How's the changes look @taylorotwell ? 😁

taylorotwell commented:
The "cute" naming part of me wants to call it something like: `$response = Gate::papers('delete', $invoice);`

Like checking / inspecting your identity papers at a border crossing, etc.

But, that is perhaps trying to be too cute.

garygreen commented:
Hmm that might be an interesting name for a package, but it's a little too quirky for my taste, doesn't seem to fit well with the current naming convention of the Gate. Just considering it as code:

```php
$papers = Gate::papers('delete', $invoice);
if ($papers->denied() && $papers->message()) {

}
```

or

```php
$access = Gate::access('delete', $invoice);
if ($access ->denied() && $access ->message()) {

}
```

To me it's patently obvious "access" is trying to perform/do a certain action. "paper" requires a deeper understanding and explanation to mean identity papers / inspection. If something requires explaining... it's probably a bit too quirky, particularly for non english speakers?

In any case, the naming can be tweaked before 5.9 is released if a more fitting name arises. Happy to do another PR to change the name at a further date if needed 😃

driesvints commented:
`Gate::access` sounds good to me as well 👍

taylorotwell commented:
Renamed to `inspect`. Also changed constructor signature of Auth Response to accept the $allowed boolean first as that feels more sensibile to me compared to have the string message be first.

@garygreen Can you please submit a PR to the master branch of the docs upgrade guide and authorization documentation pages that reflect your changes?

xuanquynh commented on src/Illuminate/Auth/Access/Response.php:
```diff
@@ -11,15 +13,33 @@ class Response
      * @return void
      */
-    public function __construct($message = null)
+    public function __construct($message, bool $allowed, $code = null)
```
The framework doesn't accept primitive type-hint, isn't?
